sympy/polys/python_powerseriesring.py

```diff
@@ -0,0 +1,384 @@
+
+    def pretty(self, series: ZZSeries) -> str:
+        coeffs, prec = series
+        return series_from_list(coeffs, prec)
+
+    def print(self, series: ZZSeries) -> str:
+        return self.pretty(series)
+
+    def equal(self, s1: ZZSeries, s2: ZZSeries) -> bool:
+        """Check if two power series are equal."""
+
+        coeffs1, prec1 = s1
+        coeffs2, prec2 = s2
+
+        return (prec1 == prec2 and
+                coeffs1 == coeffs2)
+
+    def negative(self, s: ZZSeries) -> ZZSeries:
+        """Return the negative of a power series."""
+
+        coeffs, prec = s
+
+        neg_coeffs = [-c for c in coeffs]
+        return neg_coeffs, prec
+
+    def add(self, s1: ZZSeries, s2: ZZSeries) -> ZZSeries:
+        """Add two power series."""
+        coeffs1, prec1 = s1
+        coeffs2, prec2 = s2
+
+        # case where both are DUPs
+        if not prec1 and not prec2:
+            coeffs1 = coeffs1[:self.prec]
+            coeffs2 = coeffs2[:self.prec]
+
+            series = dup_add(coeffs1[::-1], coeffs2[::-1], self.domain)
+            series = series[::-1][:self.prec]
+            if len(series) < self.prec:
+                return series, None
+            return series, self.prec
+
+        coeffs1, coeffs2, prec = _unify_prec(s1, s2)
+        series = dup_add(coeffs1[::-1], coeffs2[::-1], self.domain)
+        series = series[::-1]
+
+        return series, prec
+
+
+    def subtract(self, s1: ZZSeries, s2: ZZSeries) -> ZZSeries:
+        """Subtract two power series."""
+        coeffs1, prec1 = s1
+        coeffs2, prec2 = s2
+
+        # case where both are DUPs
+        if not prec1 and not prec2:
+            coeffs1 = coeffs1[:self.prec]
+            coeffs2 = coeffs2[:self.prec]
+
+            series = dup_sub(coeffs1[::-1], coeffs2[::-1], self.domain)
+            series = series[::-1][:self.prec]
+            if len(series) < self.prec:
+                return series, None
+            return series, self.prec
+
+        coeffs1, coeffs2, prec = _unify_prec(s1, s2)
+        series = dup_sub(coeffs1[::-1], coeffs2[::-1], self.domain)
+        series = series[::-1]
+
+        return series, prec
+
+    def multiply(self, s1: ZZSeries, s2: ZZSeries) -> ZZSeries:
+        """Multiply two power series."""
+        coeffs1, prec1 = s1
+        coeffs2, prec2 = s2
+
+        if not prec1 and not prec2:
+            coeffs1 = coeffs1[:self.prec]
+            coeffs2 = coeffs2[:self.prec]
+            series = dup_mul(coeffs1[::-1], coeffs2[::-1], self.domain)
+            series = series[::-1][:self.prec]
+            if len(series) < self.prec:
+                return series, None
+            return series, self.prec
+
+        if not prec1:
+            min_prec = prec2
+        elif not prec2:
+            min_prec = prec1
+        else:
+            low_expv1 = _first_nonzero_index(coeffs1)
+            low_expv2 = _first_nonzero_index(coeffs2)
+            min_prec = min(prec1+low_expv2, prec2+low_expv1)
```

Comment:
The fact that this needs `_first_nonzero_index` makes me wonder if it is a good idea to do this. In some domains it is difficult to check for zero e.g. RR or EX.